datasets/name_list.py

```diff
@@ -56,24 +60,47 @@ def generate_examples(self, num_examples):
                 evaluation_fn=self.evaluate_correct,
                 number_of_questions=self.count_questions(is_question),
                 is_question=is_question,
+                reset_message=self.reset_message
             )
             examples.append(example)
         return examples
 
     def evaluate_correct(
         self, questions: List[str], responses: List[str], expected_answers: List[str]
     ) -> Tuple[int, int, List[str]]:
-        not_found = []
-        for name in expected_answers:
-            if not name.lower() in responses[-1].lower():
-                not_found.append(name)
-        count = len(expected_answers) - len(not_found)
-        if len(not_found) == 0:
-            reasoning = "All expected names were found in the response."
-        else:
-            not_found = ", ".join(not_found)
-            reasoning = f"{count} names out of {len(expected_answers)} were found. Names {not_found} were not in the response."
-        return count, len(expected_answers), [reasoning]
+        try:
+            reasoning = []
+            answer_items = [n.lower() for n in sanitize_and_parse_json(responses[0])]
+            correct = 0
+            penalties = 0
+            not_found = []
+            lowered_expected = [n.lower() for n in expected_answers]
+
+            # Check answer -> expected
+            for name in answer_items:
+                if name in lowered_expected:
+                    correct += 1
+                else:
+                    penalties += 1
+                    reasoning.append(f"Name: {name} not expected.")
+
+            # Check expected -> answer
+            for name in lowered_expected:
+                if name not in answer_items:
+                    not_found.append(name)
+
+            score = correct - penalties
+            if len(not_found) > 0:
+                reasoning.append(f"Names {', '.join(not_found)} were not in the response.")
+            else:
+                reasoning.append("All expected names were found in the response.")
+
+            return score, len(expected_answers), reasoning
+
+        except (TypeError, ValueError, JSONDecodeError):
+            logging.exception(f"Response not in correct format")
```

Comment:
An error can come from using the otherwise valid JSON object as well. For example if the agent returns:  `{"list": []}` instead of just `[]`. I will see what I can do to clean it up though.